ocaml/xapi/xapi_session.ml

```diff
@@ -732,6 +732,61 @@ module Caching = struct
     ; subject_name: string
     ; rbac_permissions: string list
   }
+
+  module type EXTERNAL_AUTH_CACHE =
+    Helpers.AuthenticationCache.S
+      with type user = string
+       and type password = string
+       and type session = external_auth_result
+
+  let create_salt =
+    let module MRNG = Mirage_crypto_rng in
+    let g = MRNG.Fortuna.create () in
+    let initial_seed =
+      let r = Random.int64 Int64.max_int in
```

Comment:
Is this an acceptable way of creating a seed? I remember @edwintorok had ideas on how to fetch randomness from the system.

Reply:
Initializing the RNG is done with:
```
let () = Mirage_crypto_rng_unix.initialize (module Mirage_crypto_rng.Fortuna)
```

According to the docs.

The source code comments also say:
```
For an initial seed from \
   getrandom(), execute \
   `Mirage_crypto_rng_unix.initialize (module Mirage_crypto_rng.Fortuna)`. \
   You can use `Mirage_crypto_rng.accumulate` and `Mirage_crypto_rng.reseed` \
   to reseed the RNG manually."
```

So you only need to call reseed if you want to accumulate additional entropy, otherwise you can keep the default seed.

> Although this module exposes a more fine-grained interface, e.g. allowing manual seeding of generators, this is intended either for implementing entropy-harvesting modules, or very specialized purposes. Users of this library should almost certainly use one of the above entropy libraries, and avoid manually managing the generator seeding.

And don't create your own 'g' as the documentation says, but use the default one:
> In the general case, users should not maintain their local instances of [g](https://ocaml.org/p/mirage-crypto-rng/latest/doc/Mirage_crypto_rng/index.html#type-g). All of the generators in a process have to compete for entropy, and it is likely that the overall result will have lower effective unpredictability.
> The recommended way to use these functions is either to accept an optional generator and pass it down, or to ignore the generator altogether, as illustrated in the [examples](https://ocaml.org/p/mirage-crypto-rng/latest/doc/Mirage_crypto_rng/index.html#rng_examples).

So I think we need to delete all the reseeding code, and just call 'initialize' (once on startup, e.g. using `let ()`).
Then `generate` *without* a `g` parameter will do the right thing.

Reply:
(the mirage crypto RNG is more complicated than it needs to be because it wants to give the application/libraries low-level control over entropy and seeding, however for most use cases you don't need that, and initializing the RNG one globally is all that is needed. It will use an appropriate entropy source based on what is available on the system.

Reply:
Looking at the code it uses all of: RDSEED when available, Whirlwind CPU race entropy, `getrandom` syscall when available. A failure or absence of any of these RNG sources won't leave you unseeded, the whirlwind one is always present. 
In the past it also used `/dev/urandom`, but that got switched to the `getrandom` syscall, which have been present since  Linux 3.17, glibc 2.25.